allennlp/data/vocabulary.py

```diff
@@ -406,6 +407,112 @@ def from_params(cls, params: Params, instances: Iterable['adi.Instance'] = None)
+        It takes same parameters as Vocabulary initializer. The token2index
+        and indextotoken mappings of calling vocabulary will be retained.
+        It is an inplace operation so None will be returned.
+        """
+        if not isinstance(max_vocab_size, dict):
+            int_max_vocab_size = max_vocab_size
+            max_vocab_size = defaultdict(lambda: int_max_vocab_size)  # type: ignore
+        min_count = min_count or {}
+        pretrained_files = pretrained_files or {}
+
+        if counter is not None:
+            # Make sure vocabulary extension is safe.
+            # Common named namespaces must have same setting for padded=True/False
+            tob_added_namespaces = list(counter.keys())
+            existing_namespaces = self.get_all_namespaces()
+            for namespace in tob_added_namespaces:
+                if namespace in existing_namespaces:
+                    existing_namespace_padded = namespace in self._non_padded_namespaces
+                    tob_added_namespace_padded = namespace in non_padded_namespaces
+                    if existing_namespace_padded != tob_added_namespace_padded:
+                        raise ConfigurationError("Common namespace {} has conflicting ".format(namespace)+
+                                                 "setting of padded = True/False. "+
+                                                 "Hence extension cannot be done.")
+            # Add new non-padded namespaces for extension
+            self.add_non_padded_namespaces(non_padded_namespaces)
+
+            for namespace in counter:
+                if namespace in pretrained_files:
+                    pretrained_list = _read_pretrained_tokens(pretrained_files[namespace])
+                else:
+                    pretrained_list = None
+                token_counts = list(counter[namespace].items())
+                token_counts.sort(key=lambda x: x[1], reverse=True)
+                max_vocab = max_vocab_size[namespace]
+                if max_vocab:
+                    token_counts = token_counts[:max_vocab]
+                for token, count in token_counts:
+                    if pretrained_list is not None:
+                        if only_include_pretrained_words:
+                            if token in pretrained_list and count >= min_count.get(namespace, 1):
+                                self.add_token_to_namespace(token, namespace)
+                        elif token in pretrained_list or count >= min_count.get(namespace, 1):
+                            self.add_token_to_namespace(token, namespace)
+                    elif count >= min_count.get(namespace, 1):
+                        self.add_token_to_namespace(token, namespace)
+
+        if tokens_to_add:
+            for namespace, tokens in tokens_to_add.items():
+                for token in tokens:
+                    self.add_token_to_namespace(token, namespace)
```

Comment:
isn't most of this code (lines 446 - 469) duplicated verbatim from `Vocabulary.__init__`? can we factor it out?

Reply:
Yes I have done exactly that. `extend` method and `__init__` have exact same arguments. Even functionally they are overlapping: `__init__` makes and empty vocabulary and then extends it with `extend`. Then we also have `exend_from_instances` which wraps `extend` method for it's purpose. Hence `extend` function is basically the factored out code between `Vocab.__init__` and `extend_from_instances`.  `__init__` function is just 4 lines of code now.

Reply:
oops, missed that, thanks!